core/src/test/scala/org/apache/spark/scheduler/EventLoggingListenerSuite.scala

```diff
@@ -251,6 +261,233 @@ class EventLoggingListenerSuite extends SparkFunSuite with LocalSparkContext wit
+      createExecutorMetricsUpdateEvent(1,
+        new ExecutorMetrics(20L, 2000L, 40L, 50L, 0L, 40L, 10L, 90L, 10L, 50L, 0L)),
+      createExecutorMetricsUpdateEvent(2,
+        new ExecutorMetrics(20L, 3500L, 50L, 15L, 0L, 10L, 10L, 35L, 10L, 80L, 0L)),
+      createStageSubmittedEvent(1),
+      createExecutorMetricsUpdateEvent(1,
+        new ExecutorMetrics(25L, 5000L, 30L, 50L, 20L, 30L, 10L, 80L, 30L, 50L, 0L)),
+      createExecutorMetricsUpdateEvent(2,
+        new ExecutorMetrics(25L, 7000L, 70L, 50L, 20L, 0L, 10L, 50L, 30L, 10L, 40L)),
+      createStageCompletedEvent(0),
+      createExecutorMetricsUpdateEvent(1,
+        new ExecutorMetrics(30L, 6000L, 70L, 20L, 30L, 10L, 0L, 30L, 30L, 30L, 0L)),
+      createExecutorMetricsUpdateEvent(2,
+        new ExecutorMetrics(30L, 5500L, 30L, 20L, 40L, 10L, 0L, 30L, 40L, 40L, 20L)),
+      createExecutorMetricsUpdateEvent(1,
+        new ExecutorMetrics(35L, 7000L, 70L, 5L, 25L, 60L, 30L, 65L, 55L, 30L, 0L)),
+      createExecutorMetricsUpdateEvent(2,
+        new ExecutorMetrics(35L, 5500L, 40L, 25L, 30L, 10L, 30L, 35L, 60L, 0L, 20L)),
+      createExecutorMetricsUpdateEvent(1,
+        new ExecutorMetrics(40L, 5500L, 70L, 15L, 20L, 55L, 20L, 70L, 40L, 20L, 0L)),
+      createExecutorRemovedEvent(1),
+      createExecutorMetricsUpdateEvent(2,
+        new ExecutorMetrics(40L, 4000L, 20L, 25L, 30L, 10L, 30L, 35L, 60L, 0L, 0L)),
+      createStageCompletedEvent(1),
+      SparkListenerApplicationEnd(1000L))
+
+    // play the events for the event logger
+    eventLogger.start()
+    listenerBus.start(Mockito.mock(classOf[SparkContext]), Mockito.mock(classOf[MetricsSystem]))
+    listenerBus.addToEventLogQueue(eventLogger)
+    events.foreach(event => listenerBus.post(event))
+    listenerBus.stop()
+    eventLogger.stop()
+
+    // Verify the log file contains the expected events.
+    // Posted events should be logged, except for ExecutorMetricsUpdate events -- these
+    // are consolidated, and the peak values for each stage are logged at stage end.
+    val logData = EventLoggingListener.openEventLog(new Path(eventLogger.logPath), fileSystem)
+    try {
+      val lines = readLines(logData)
+      val logStart = SparkListenerLogStart(SPARK_VERSION)
+      assert(lines.size === 14)
+      assert(lines(0).contains("SparkListenerLogStart"))
+      assert(lines(1).contains("SparkListenerApplicationStart"))
+      assert(JsonProtocol.sparkEventFromJson(parse(lines(0))) === logStart)
+      var i = 1
+      events.foreach {event =>
+        event match {
+          case metricsUpdate: SparkListenerExecutorMetricsUpdate =>
+          case stageCompleted: SparkListenerStageCompleted =>
+            for (j <- 1 to 2) {
+              checkExecutorMetricsUpdate(lines(i), stageCompleted.stageInfo.stageId,
+                expectedMetricsEvents)
+                i += 1
+             }
+            checkEvent(lines(i), event)
+            i += 1
```

Comment:
I found this pretty confusing at first.  I suggest renaming `i` to `logIdx` and including a comment about the `j` loop.  Also we tend to use `(1 to 2).foreach`.  eg.

```scala
// just before the SparkListenerStageCompleted gets logged, we expect to get a 
// SparkListenerExecutorMetricsUpdate for each executor
(1 to 2).foreach { _ =>
  checkExecutorMetricsUpdate(lines(logIdx), stageCompleted.stageInfo.stageId,
    expectedMetricsEvents)
  logIdx += 1
}
// also check that we get the expected SparkListenerStageCompleted
checkEvent(lines(logIdx), event)
logIdx += 1
```

Reply:
Changed for both.